src/core/jit/synth/synth.rs

```diff
@@ -129,25 +131,42 @@ impl<'a, Value: JsonLike<'a> + Clone + 'a> Synth<Value> {
         node: &'a Field<Nested<Value>, Value>,
         parent: &'a Value,
         data_path: &DataPath,
-    ) -> Result<Value, Positioned<Error>> {
+    ) -> Result<Value, LocationError<Error>> {
         let include = self.include(node);
-        if include && node.is_scalar {
+        if parent.is_null() {
+            if node.type_of.is_nullable() {
+                Ok(Value::null())
+            } else {
+                Err(ValidationError::ValueRequired.into())
+                    .map_err(|e| self.to_location_error(e, node))
+            }
+        } else if include && self.plan.field_is_scalar(node) {
             let scalar = scalar::get_scalar(node.type_of.name());
-
             // TODO: add validation for input type as well. But input types are not checked
             // by async_graphql anyway so it should be done after replacing
             // default engine with JIT
             if (scalar.validate())(parent) {
                 Ok(parent.clone())
             } else {
-                Err(Positioned {
-                    pos: node.pos,
-                    node: ValidationError::ScalarInvalid {
-                        type_of: node.type_of.name().to_string(),
-                        path: node.name.clone(),
-                    }
-                    .into(),
-                })
+                Err(
+                    ValidationError::ScalarInvalid { type_of: node.type_of.name().to_string() }
+                        .into(),
+                )
+                .map_err(|e| self.to_location_error(e, node))
+            }
+        } else if self.plan.field_is_enum(node) {
+            if parent
+                .as_str()
+                .map(|v| self.plan.field_validate_enum_value(node, v))
+                .unwrap_or(false)
+            {
+                Ok(parent.clone())
+            } else {
+                Err(
+                    ValidationError::EnumInvalid { type_of: node.type_of.name().to_string() }
+                        .into(),
+                )
+                .map_err(|e| self.to_location_error(e, node))
```

Comment:
we can have single `map_err` call to the whole if-else block that way reducing duplication in every branch